File: packages/shield-py/shield/anonymiser.py
"""
Name anonymisation: replacement, ID token generation, row shuffling.
Pure functions, stdlib only.
"""

import json
import re
from pathlib import Path
from typing import Any, Callable, Dict, List, Optional, Set, Tuple

from .prng import shuffle as prng_shuffle

# Load name bank
_DATA_DIR = Path(__file__).parent / 'data'
with open(_DATA_DIR / 'names.json', 'r', encoding='utf-8') as _f:
    _names = json.load(_f)


def build_blocked_tokens(
    rows: List[List[str]],
    name_col_index: int,
    flags: Optional[List[Any]] = None,
    staff_names: Optional[List[str]] = None,
) -> Set[str]:
    """Build a set of blocked tokens from student names, flags, and staff names.

    All tokens are lowercased for case-insensitive matching.

    Args:
        rows: Data rows (no header).
        name_col_index: Index of the name column.
        flags: Flagged tokens from scanner.
        staff_names: Staff names to block.

    Returns:
        Set of blocked tokens (lowercased).
    """
    blocked: Set[str] = set()

    # Add all real student first names and surnames individually
    for row in rows:
        full_name = row[name_col_index]
        if not isinstance(full_name, str):
            continue
        parts = full_name.strip().split()
        for part in parts:
            if len(part) > 0:
                blocked.add(part.lower())

    # Add all flagged tokens
    if flags:
        for flag in flags:
            if isinstance(flag, str):
                token = flag
            elif isinstance(flag, dict):
                token = flag.get('value') or flag.get('token')
            else:
                token = None
            if token:
                blocked.add(token.lower())

    # Add all staff names (split into individual parts)
    if staff_names:
        for name in staff_names:
            parts = name.strip().split()
            for part in parts:
                if len(part) > 0:
                    blocked.add(part.lower())

    return blocked


def _select_name(
    name_list: List[str],
    blocked_tokens: Set[str],
    used_names: Set[str],
    prng: Callable[[], float],
) -> str:
    """Select a name from a name list using the PRNG, avoiding blocked tokens.

    Consumes one PRNG call per attempt.

    Args:
        name_list: Array of names to choose from.
        blocked_tokens: Tokens to avoid (lowercased).
        used_names: Already-used names to avoid reuse (lowercased).
        prng: PRNG function.

    Returns:
        Selected name.

    Raises:
        RuntimeError: If name bank is exhausted.
    """
    max_attempts = len(name_list) * 3
    for _ in range(max_attempts):
        index = int(prng() * len(name_list))
        name = name_list[index]
        name_lower = name.lower()
        if name_lower not in blocked_tokens and name_lower not in used_names:
            return name

    # Fallback: exhaustive search
    for name in name_list:
        name_lower = name.lower()
        if name_lower not in blocked_tokens and name_lower not in used_names:
            return name

    raise RuntimeError('Unable to find a non-blocked, unused name. Name bank exhausted.')


def _get_first_name_list(
    gender: Optional[str],
    gender_mode: str,
) -> List[str]:
    """Determine which first-name list to use based on gender.

    Args:
        gender: Gender value from the row.
        gender_mode: 'aware' or 'neutral'.

    Returns:
        List of first names to choose from.
    """
    if gender_mode == 'neutral':
        return _names['neutral']

    if gender:
        g = gender.lower().strip()
        if g in ('m', 'male'):
            return _names['male']
        if g in ('f', 'female'):
            return _names['female']
        if g in ('x', 'non-binary', 'nonbinary', 'other'):
            return _names['neutral']

    # Default: combine all lists
    return _names['male'] + _names['female'] + _names['neutral']


def anonymise_names(
    rows: List[List[str]],
    name_col_index: int,
    prng: Callable[[], float],
    options: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Anonymise names in data rows, replacing real names with fake names and ID tokens.

    PRNG consumption order: one call for first name index, one for surname index, per student.

    Args:
        rows: Data rows (no header).
        name_col_index: Index of the name column.
        prng: PRNG function.
        options: Options dict with gender_col_index, gender_mode, blocked_tokens.

    Returns:
        Dict with 'rows' (anonymised) and 'mapping'.
    """
    if options is None:
        options = {}

    gender_col_index = options.get('gender_col_index')
    gender_mode = options.get('gender_mode', 'aware')
    blocked_tokens: Set[str] = options.get('blocked_tokens', set())

    mapping = {
        'ids': {},
        'names': {},
        'nameToId': {},
        'idToFake': {},
    }

    used_first_names: Set[str] = set()
    used_surnames: Set[str] = set()
    new_rows = []

    for i, row in enumerate(rows):
        row = list(row)  # Copy
        real_name = row[name_col_index]
        id_num = i + 1
        id_str = f'0{id_num}' if id_num < 10 else str(id_num)
        id_token = f'S{id_str}'

        # Determine gender for name selection
        gender = row[gender_col_index] if gender_col_index is not None else None
        first_name_list = _get_first_name_list(gender, gender_mode)

        # Select first name (one PRNG call)
        first_name = _select_name(first_name_list, blocked_tokens, used_first_names, prng)
        used_first_names.add(first_name.lower())

        # Select surname (one PRNG call)
        surname = _select_name(_names['surnames'], blocked_tokens, used_surnames, prng)
        used_surnames.add(surname.lower())

        fake_name = f'{first_name} {surname}'

        # Store mapping
        mapping['ids'][id_token] = real_name
        mapping['names'][fake_name] = real_name
        mapping['nameToId'][fake_name] = id_token
        mapping['idToFake'][id_token] = fake_name

        # Replace name column with "FakeName [SXX]" format
        row[name_col_index] = f'{fake_name} [{id_token}]'
        new_rows.append(row)

    return {'rows': new_rows, 'mapping': mapping}


def shuffle_rows(
    rows: List[List[Any]],
    prng: Callable[[], float],
) -> List[List[Any]]:
    """Shuffle rows using Fisher-Yates algorithm with the seeded PRNG.

    Returns a new list (does not mutate the input).

    Args:
        rows: Rows to shuffle.
        prng: PRNG function.

    Returns:
        New list with rows in shuffled order.
    """
    copy = list(rows)
    prng_shuffle(copy, prng)
    return copy

File: packages/shield/src/anonymiser.js
import { shuffle } from './prng.js';
import names from './data/names.json' with { type: 'json' };

/**
 * Build a set of blocked tokens from student names, flags, and staff names.
 * All tokens are lowercased for case-insensitive matching.
 *
 * @param {Array<Array<string>>} rows - Data rows (no header)
 * @param {number} nameColIndex - Index of the name column
 * @param {Array<string|{value: string}|{token: string}>} [flags] - Flagged tokens from scanner
 * @param {string[]} [staffNames] - Staff names to block
 * @returns {Set<string>} Set of blocked tokens (lowercased)
 */
export function buildBlockedTokens(rows, nameColIndex, flags = [], staffNames = []) {
  const blocked = new Set();

  // Add all real student first names and surnames individually
  for (const row of rows) {
    const fullName = row[nameColIndex];
    if (typeof fullName !== 'string') continue;
    const parts = fullName.trim().split(/\s+/);
    for (const part of parts) {
      if (part.length > 0) {
        blocked.add(part.toLowerCase());
      }
    }
  }

  // Add all flagged tokens
  if (flags) {
    for (const flag of flags) {
      const token = typeof flag === 'string' ? flag : (flag.value || flag.token);
      if (token) {
        blocked.add(token.toLowerCase());
      }
    }
  }

  // Add all staff names (split into individual parts)
  if (staffNames) {
    for (const name of staffNames) {
      const parts = name.trim().split(/\s+/);
      for (const part of parts) {
        if (part.length > 0) {
          blocked.add(part.toLowerCase());
        }
      }
    }
  }

  return blocked;
}

/**
 * Select a name from a name list using the PRNG, avoiding blocked tokens.
 * Consumes one PRNG call per attempt.
 *
 * @param {string[]} nameList - Array of names to choose from
 * @param {Set<string>} blockedTokens - Tokens to avoid (lowercased)
 * @param {Set<string>} usedNames - Already-used names to avoid reuse (lowercased)
 * @param {function(): number} prng - PRNG function
 * @returns {string} Selected name
 */
function selectName(nameList, blockedTokens, usedNames, prng) {
  const maxAttempts = nameList.length * 3;
  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    const index = Math.floor(prng() * nameList.length);
    const name = nameList[index];
    const nameLower = name.toLowerCase();
    if (!blockedTokens.has(nameLower) && !usedNames.has(nameLower)) {
      return name;
    }
  }
  // Fallback: exhaustive search for any unblocked, unused name
  for (const name of nameList) {
    const nameLower = name.toLowerCase();
    if (!blockedTokens.has(nameLower) && !usedNames.has(nameLower)) {
      return name;
    }
  }
  throw new Error('Unable to find a non-blocked, unused name. Name bank exhausted.');
}

/**
 * Determine which first-name list to use based on gender.
 *
 * @param {string|undefined} gender - Gender value from the row
 * @param {string} genderMode - 'aware' or 'neutral'
 * @returns {string[]} Array of first names to choose from
 */
function getFirstNameList(gender, genderMode) {
  if (genderMode === 'neutral') {
    return names.neutral;
  }
  if (gender) {
    const g = gender.toLowerCase().trim();
    if (g === 'm' || g === 'male') return names.male;
    if (g === 'f' || g === 'female') return names.female;
    if (g === 'x' || g === 'non-binary' || g === 'nonbinary' || g === 'other') return names.neutral;
  }
  // Default: combine all lists
  return [...names.male, ...names.female, ...names.neutral];
}

/**
 * Normalise a name from "Surname, Firstname" to "Firstname Surname" format.
 * If the name doesn't contain a comma, returns it unchanged.
 *
 * @param {string} name
 * @returns {string}
 */
function normaliseNameOrder(name) {
  if (typeof name !== 'string') return name;
  const commaIndex = name.indexOf(',');
  if (commaIndex === -1) return name;
  const before = name.slice(0, commaIndex).trim();
  const after = name.slice(commaIndex + 1).trim();
  if (before && after) return `${after} ${before}`;
  return name;
}

/**
 * Anonymise names in data rows, replacing real names with fake names and ID tokens.
 *
 * PRNG consumption order: one call for first name index, one for surname index, per student.
 *
 * @param {Array<Array<string>>} rows - Data rows (no header)
 * @param {number} nameColIndex - Index of the name column
 * @param {function(): number} prng - PRNG function
 * @param {object} [options] - Options
 * @param {number} [options.genderColIndex] - Index of gender column for gender-aware naming
 * @param {string} [options.genderMode='aware'] - 'aware' or 'neutral'
 * @param {Set<string>} [options.blockedTokens] - Set of blocked tokens (lowercased)
 * @returns {{ rows: Array<Array<string>>, mapping: object }} Anonymised rows and mapping
 */
export function anonymiseNames(rows, nameColIndex, prng, options = {}) {
  const {
    genderColIndex,
    genderMode = 'aware',
    blockedTokens = new Set()
  } = options;

  const mapping = {
    ids: {},       // { 'S01': 'Real Name' }
    names: {},     // { 'Fake Name': 'Real Name' }
    nameToId: {},  // { 'Fake Name': 'S01' }
    idToFake: {}   // { 'S01': 'Fake Name' }
  };

  const usedFirstNames = new Set();
  const usedSurnames = new Set();
  const newRows = [];

  for (let i = 0; i < rows.length; i++) {
    const row = [...rows[i]];
    const rawName = row[nameColIndex];
    const realName = normaliseNameOrder(rawName);
    const idNum = i + 1;
    const idStr = idNum < 10 ? `0${idNum}` : `${idNum}`;
    const idToken = `S${idStr}`;

    // Determine gender for name selection
    const gender = genderColIndex !== undefined ? row[genderColIndex] : undefined;
    const firstNameList = getFirstNameList(gender, genderMode);

    // Select first name (one PRNG call)
    const firstName = selectName(firstNameList, blockedTokens, usedFirstNames, prng);
    usedFirstNames.add(firstName.toLowerCase());

    // Select surname (one PRNG call)
    const surname = selectName(names.surnames, blockedTokens, usedSurnames, prng);
    usedSurnames.add(surname.toLowerCase());

    const fakeName = `${firstName} ${surname}`;

    // Store mapping — always "Firstname Surname" order for clean report output
    mapping.ids[idToken] = realName;
    mapping.names[fakeName] = realName;
    mapping.nameToId[fakeName] = idToken;
    mapping.idToFake[idToken] = fakeName;

    // Replace name column with "FakeName [SXX]" format
    row[nameColIndex] = `${fakeName} [${idToken}]`;
    newRows.push(row);
  }

  return { rows: newRows, mapping };
}

/**
 * Shuffle rows using Fisher-Yates algorithm with the seeded PRNG.
 * Returns a new array (does not mutate the input).
 *
 * @param {Array<Array>} rows - Rows to shuffle
 * @param {function(): number} prng - PRNG function
 * @returns {Array<Array>} New array with rows in shuffled order
 */
export function shuffleRows(rows, prng) {
  const copy = [...rows];
  shuffle(copy, prng);
  return copy;
}
